AMD Strix Halo SAM3 Semantic Navigation (#929)
ros-planning/navigation.ros.org: tutorials/docs/navigation2_with_sam3_semantic_segmentation.rst

The semantic costmap layer takes those three pieces: the mask, the labels, and the registered pointcloud.
It projects each labeled mask pixel into 3D using the pointcloud and accumulating observations per costmap tile to set its class type.
Internally, the ``semantic_segmentation_layer`` maintains a *tile map* to store these observations over a sliding window used to populate the costmap.

Tutorial Steps
==============

0 - Clone the repository
------------------------

.. code-block:: bash

   cd <your_ros2_ws>/src
   git clone --recursive https://github.com/open-navigation/opennav_amd_semantic_navigation.git

If you already cloned without ``--recursive`` (to get the costmap layer submodule):

.. code-block:: bash

   cd opennav_amd_semantic_navigation
   git submodule update --init --recursive

1 - Install ROCm and Python dependencies
----------------------------------------

The ``opennav_sam3_inference`` package ships a one-shot setup script that installs the pinned ROCm 7.2 stack, a patched MIGraphX 2.15, an ``opennav-sam3-inference`` conda environment with ROCm-nightly PyTorch + ``onnxruntime-migraphx``, and (optionally) the model weights

> That sounds like alot, but basically it sets up the right versions of AMD's software, AI optimization libraries, and PyTorch so that everyone plays nicely and there is no dependency hell.

Author: ```suggestion
That sounds like a bunch, but basically it sets up the right versions of AMD's software, AI optimization libraries, and PyTorch so that everyone plays nicely and there is no dependency hell.
```

- ``score_threshold`` / ``mask_threshold`` - confidence thresholds for whole detections and for per-pixel binarization.
- ``redetect_interval_ms`` - run full SAM3 detection every N milliseconds; track-only on the rest. ``0`` redetects every frame; larger values cut compute at the cost of catching new objects more slowly.
- ``max_objects_per_prompt`` - cap on simultaneously tracked instances per prompt. Excess (lowest score) are evicted so the tracker stops propagating them.

Services are exposed to enable/disable or change the prompts and class IDs at run-time.

A few prompt-design lessons worth keeping in mind:

Prompts can be long phrases that group multiple physical objects under one ID - for example ``"car, bus, plane, or motorcycle"`` is *one* prompt.
Fewer prompts run faster as SAM3 will do processing over each prompt given. Try to consolidate as much as possible or use liberal use of the prompt change service.

5 - Build the ROS 2 workspace
-----------------------------

Build the SAM3 inference package inside the conda environment so the SAM3 node picks up the right Python.

> After built, you do not need to activate the conda enviornment again - launching the node will automatically handle the environment correctly!

Author: ```suggestion
After built, you do not need to activate the conda environment again - launching the node will automatically handle the environment correctly!
```

   </p>

SAM3 does an amazing job with no fine-tuning across a huge variety of terrain types and environments, which is a game-changer for navigation in unstructured environments.
The accuracy, sharpness, and consistency of the masks is a huge step up from traditional semantic segmentation models, and the text promptability means you can segment out whatever classes are relevant to your application on Day 1!

Difficult Small Obstacles
-------------------------

The previous demos showcase terrain.
The other side of foundation-model perception is that you can ask SAM3 about specific things like cables on the floor, a puddle, a piece of debris, a low-profile pallet without training a detector for each one.
I hope these provide some motivational examples of use-cases of SAM3 to detect and avoid obstacles (or adjust robot behavior) in situations that have been traditionally difficult for robots :-)

.. image:: images/semantic_strix_halo/examples.gif
    :width: 90%
    :align: center
    :alt: SAM3 detecting cables, spills, debris, pallets, and other low-profile obstacles across various test images

> Happy segmenting!

Author: ```suggestion
Happy segmenting!

```
